crates/fuel-core/src/schema/tx/assemble_tx.rs

```diff
@@ -0,0 +1,879 @@
+            .tx_params()
+            .max_gas_per_tx();
+        let gas_price_factor = fee_params.gas_price_factor();
+        let fee_payer_account = self.fee_payer_account.clone();
+
+        let mut total_base_asset = 0u64;
+
+        for input in self.tx.inputs() {
+            let Some(amount) = input.amount() else {
+                continue;
+            };
+            let Some(asset_id) = input.asset_id(&base_asset_id) else {
+                continue;
+            };
+            let Some(owner) = input.input_owner() else {
+                continue;
+            };
+
+            if asset_id == &base_asset_id && &fee_payer_account.owner() == owner {
+                total_base_asset =
+                    total_base_asset.checked_add(amount).ok_or(anyhow::anyhow!(
+                        "The total base asset amount used by the transaction is too big"
+                    ))?;
+            }
+        }
+
+        loop {
+            let remaining_input_slots = self.remaining_input_slots()?;
+            let max_gas = self.tx.max_gas(&gas_costs, &fee_params);
+            let max_gas_with_reserve = max_gas.saturating_add(self.arguments.reserve_gas);
+
+            let final_gas = max_gas_with_reserve.min(max_gas_per_tx);
+            let final_fee =
+                gas_to_fee(final_gas, self.arguments.gas_price, gas_price_factor);
+            let final_fee = u64::try_from(final_fee)
+                .map_err(|_| {
+                    anyhow::anyhow!("The final fee is too big to fit into `u64`")
+                })?
+                .saturating_add(self.tx.tip());
+
+            let need_to_cover = final_fee.saturating_add(self.base_asset_reserved);
+
+            if need_to_cover <= total_base_asset {
+                break;
+            }
+
+            let how_much_to_add = need_to_cover.saturating_sub(total_base_asset);
+            let coins = self
+                .arguments
+                .coins(
+                    fee_payer_account.owner(),
+                    base_asset_id,
+                    how_much_to_add,
+                    remaining_input_slots,
+                )
+                .await?;
+
+            if coins.is_empty() {
+                return Err(anyhow::anyhow!(
+                    "Unable to find any coins to pay for the fee"
+                ));
+            }
+
+            for coin in coins {
+                total_base_asset = total_base_asset.checked_add(coin.amount()).ok_or(
+                    anyhow::anyhow!(
+                        "The total base asset amount \
+                        became too big when tried to cover fee"
+                    ),
+                )?;
+                self.add_input_and_witness_and_change(&fee_payer_account, coin);
+            }
+
+            // In the case when predicates iterates over the inputs,
+            // it increases its used gas. So we need to re-estimate predicates.
+            self = self.estimate_predicates().await?;
```

Comment:
We can't do it only once at the end ? It seems very heavy operation it scares me.

Reply:
Hmm yeah it seems to me that we don't need the predicate estimations within this loop so we might as well do the re-estimation afterwards. Though good to verify with @xgreenx if we're missing something here.

Reply:
The issues that each time when we add a new coin, it is possible that it will change the gas used by predicate, causing us to find more coins, that can change predicates again, etc.

Reply:
Added protection to not do more than 10 estimation of predicates